synapseclient/models/activity.py

```diff
@@ -1 +1,345 @@
-# TODO
+import asyncio
+from dataclasses import dataclass, field
+from typing import List, Dict, Union, Tuple
+
+from synapseclient.activity import Activity as Synapse_Activity
+from opentelemetry import trace, context
+
+from typing import Optional, TYPE_CHECKING
+
+from synapseclient import Synapse
+from synapseclient.core.async_utils import otel_trace_method
+
+if TYPE_CHECKING:
+    from synapseclient.models import File, Table
+
+USED_URL_CONCRETE_TYPE = "org.sagebionetworks.repo.model.provenance.UsedURL"
+USED_ENTITY_CONCRETE_TYPE = "org.sagebionetworks.repo.model.provenance.UsedEntity"
+
+tracer = trace.get_tracer("synapseclient")
+
+
+@dataclass
+class UsedEntity:
```

Comment:
Nit: Semantics, but this is `UsedEntityReference`: https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/Reference.html.  `UsedEntity`: https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/provenance/UsedEntity.html


Did we want this class to be consistent with the API?  I know there is logic that relies on this class being this way.

Reply:
I had thought about having 2 layers for this to model exactly how the REST API represents there. However, I elected to follow what I did in the code here to remove the nested objects.

I also removed `wasExecuted` and `concreteType` anywhere from the OOP model attributes. Instead separating them out into 2 different lists on the Activity. IMO this makes it a bit easier to work with by having separate collections of items rather than a boolean that controls it.

```
    used: List[Union[UsedEntity, UsedURL]] = field(default_factory=list)
    """The entities used by this Activity."""

    executed: List[Union[UsedEntity, UsedURL]] = field(default_factory=list)
    """The entities executed by this Activity."""
```

Thoughts on how I am diverging from the REST API here?

One of the possible future downsides here is that if the synapse REST API changes in such a way that conflicts with my implementation, I'd need to go back on some of these decisions.

Reply:
This makes sense to me.  Thanks!